### PR TITLE
Finally com_mailto allow the usage of a captcha by using JForm

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

Finally com_mailto allow the usage of a captach by using JForm 

As discussed with the JSST this is now a public tracker item: cc @SniperSister @mbabker

### Testing Instructions

- apply this patch
- check that the mail to friend feature still works
- configure the captcha
- check that com_mailto now requires the captcha

### Expected result

com_mailto supports captcha's
com_mailto uses JForm

### Actual result

com_mailto don't supports captcha's
com_mailto don't uses JForm

### Documentation Changes Required

none

### CODE REVIEW

File: components/com_mailto/models/mailto.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ */
+class MailtoModelMailto extends JModelForm
+{
+	/**
+	 * Method to get the mailto form.
+	 *
+	 * The base form is loaded from XML and then an event is fired
+	 * for users plugins to extend the form with extra fields.
+	 *
+	 * @param   array    $data      An optional array of data for the form to interogate.
```
Comment (brianteeman) [inline components/com_mailto/models/mailto.php]: interrogate

**Review by brianteeman — CHANGES_REQUESTED** [review]

File: components/com_mailto/models/mailto.php (hunk 1 above)
Comment (zero-24) [reply components/com_mailto/models/mailto.php]: fixed thanks.

**Comment by brianteeman** [comment]
milestone is set to 3.8.8. but surely this is a new feature according to semver

**Comment by zero-24** [comment]
> milestone is set to 3.8.8. but surely this is a new feature according to semver

new feature? Where do you see a new feature? This is a fix for a broken and exploited attack against sites as noted here in public and in private many times. ;)

File: components/com_mailto/views/mailto/view.html.php [hunk 2]
```diff
@@ -27,62 +27,10 @@ class MailtoViewMailto extends JViewLegacy
...
-		$data = $this->getData();
+		$this->form = $this->get('Form');
 
-		if ($data === false)
-		{
-			return false;
-		}
-
-		$this->set('data', $data);
+		$this->set('link', urldecode(JFactory::getApplication()->input->get('link', '', 'BASE64')));
```
Comment (mbabker) [inline components/com_mailto/views/mailto/view.html.php]: `$this->link = ...`

**Comment by mbabker** [comment]
You technically get a new feature introducing captcha support to the form.  Otherwise it's all just internal refactoring/restructuring.

**Comment by zero-24** [comment]
> You technically get a new feature introducing captcha support to the form. 

Correct but this is just a side effect of the refactoring to JForm ;)

**Comment by brianteeman** [comment]
Adding captcha is a new feature. Semantic versioning is all about technicalities

**Comment by zero-24** [comment]
@mbabker do you agree with @brianteeman or can we move forward with that protection against spam in 3.8.8 by allowing that `feature`?

Or do we need to go the hard way in patching the xml lines for the captcha out and explain in a doc page where to add the support for that.?

**Comment by ot2sen** [comment]
Must be considered a bug at this stage. Captcha has been a core feature by ages so it must be a bug that it didn´t arrive in all contact or form interactions when first added :)

File: components/com_mailto/views/mailto/view.html.php (hunk 2 above)
Comment (zero-24) [reply components/com_mailto/views/mailto/view.html.php]: Fixed with the last commit. Thanks.

**Comment by mbabker** [comment]
To be completely honest it's one of those edge cases where you have to outweigh the user benefit with strict adherence to the standard.  At this point we're dealing with a weakness in core that could've legitimately just been committed to 3.8.8 as a security fix without this public item but we collectively decided this doesn't need to be committed with that type of "secrecy" (for a lack of a better term).  So for me, even though a standard says this strictly must trigger a minor version release, I don't think this absolutely must go into 3.9 and not be addressed in a 3.8 release.

**Comment by brianteeman** [comment]
@mbabker even if it was a security issue it would still require a version bump
@ot2sen that would be perhaps true if it was indeed added to all forms but it still hasnt been

there are a lot more urgent issues which are actually going to b useful that are being held up for 3.9

either we use semver or we dont - its not something that can be done half heartedly

**Comment by mbabker** [comment]
It's not being done half-heartedly.  The standard is the guiding line and we have to justify deviating from it (which should be an exceptional basis).  It's not "follow the standard when it's convenient".  If we followed the standard that strictly then we would have to version bump when we:

- Run `composer update` and it includes minor version increments (the changes may not all be part of our API but as part of our distributed package it does introduce new features)
- Update any JavaScript library which is a minor version release (CodeMirror being the common one)
- Merge any patch that is not strictly fixing a bug

So we already do it "half heartedly" in some ways.  Considering I can't get good answers on how to handle releases beyond 3.8.x right now the best thing I can do while I'm running releases is use my best judgment on things.  And considering the amount of marketing hype that the project likes to put onto .0 releases (and the amount of annoyance end users have if .0 releases don't have something new and shiny, see 3.4 and 3.5 reactions) I doubt anyone would want us doing "full blown" feature releases every 2-3 months as that's about the cycle we're on for something being merged that is not 100% in line with the SemVer definition of a patch release.

**Comment by brianteeman** [comment]
guess i am just frustrated that several pr of mine got delayed because of semver and the privacy plugin will be useless by the time it is released

**Comment by Quy** [comment]
The form no longer fits within the popup window. You have to scroll down to see the buttons.
The email address is not populated.

Before:
20265-before

After:
20265-after

**Comment by zero-24** [comment]
should be fixed now thanks @Quy

**Comment by brianteeman** [comment]
the last change isnt a good solution. You have hard coded the height of the popup to the height required by the core captcha irrespective of it being enabled or not. In addition the user could be using a different captcha plugin that requires more space. The modal height should be dynamic based on the content and not fixed or at least in a file that the user can override.

**Comment by zero-24** [comment]
Still not great but better than nothing.

**Comment by Quy** [comment]
Without captcha, it should be 500.

**Comment by brianteeman** [comment]
~did the JSST review the changes in this pr - there are issues (not for public explanation)~

**Comment by brianteeman** [comment]
Note the style changes because the markup is different which  effects the height

### Before PR
chrome_2018-05-01_08-50-38



### After PR
chrome_2018-05-01_08-52-22

**Comment by zero-24** [comment]
Fixed thanks. @brianteeman

**Comment by zero-24** [comment]
> Without captcha, it should be 500.

well with captcha we than have just 50 more than we can just do it 550 for all?

**Comment by zero-24** [comment]
> did the JSST review the changes in this pr - there are issues (not for public explanation)

Just to don't let this claim unanswered. We are in 2018 such a check should not be required any more (It seams to be something added in 2006). But I have just added the check back. And for the record I got an review on the code first for sure. Thanks.

**Comment by brianteeman** [comment]
This PR doesnt "allow" the use of captcha it forces it on unless you turn it off - is that the intended behaviour?

Maybe I am thick but I dont see where the mailto xml options can be accessed

**Comment by zero-24** [comment]
> This PR doesnt "allow" the use of captcha it forces it on unless you turn it off - is that the intended behaviour?

Well when you have setup a captcha joomla should use it by default like we did for the registration right?

**Comment by brianteeman** [comment]
> We are in 2018 such a check should not be required any more (It seams to be something added in 2006)

the issue addressed back then is still a valid issue today so it is still required

**Comment by zero-24** [comment]
> the issue addressed back then is still a valid issue today so it is still required

As i said this check was added back. ;) And even improved :P

**Comment by brianteeman** [comment]
> Maybe I am thick but I dont see where the mailto xml options can be accessed

**Comment by zero-24** [comment]
> Maybe I am thick but I dont see where the mailto xml options can be accessed

As there are no options for com_mailto itself no where :P You can enable and disable the mailto link in com_content / global / menu.

**Comment by brianteeman** [comment]
so why create https://github.com/joomla/joomla-cms/pull/20265/files#diff-01460342ce71aca88928e9c8be184f21

**Comment by mbabker** [comment]
To render a JForm instance on the frontend it still uses those XML definitions.

**Comment by mbabker** [comment]
Plus it defines some rules that are used in form validation.  It's not strictly for backend configuration.

**Comment by brianteeman** [comment]
see told you I was thick. I guess it was an error then that we didnt use an xml before

**Comment by mbabker** [comment]
Well when you have a half-million line code base and few people looking over and maintaining all aspects of it... 😆

**Comment by zero-24** [comment]
> see told you I was thick. I guess it was an error then that we didnt use an xml before

Yes. We didn't needed that xml before as we just did plain html without using jform now with jform the xml is used for the things Michael posted configuration & validation settings. ;)

**Comment by Quy** [comment]
In Chrome without captcha, there is a vertical scrollbar. 
20265-chrome

File: components/com_mailto/views/mailto/tmpl/default.php [hunk 3]
```diff
@@ -41,44 +24,35 @@
...
-		<p>
-			<button type="button" class="button" onclick="return Joomla.submitbutton('send');">
-				<?php echo JText::_('COM_MAILTO_SEND'); ?>
-			</button>
-			<button type="button" class="button" onclick="window.close();return false;">
-				<?php echo JText::_('COM_MAILTO_CANCEL'); ?>
-			</button>
-		</p>
+	<form action="<?php echo JRoute::_('index.php?option=com_mailto&task=send'); ?>" method="post" class="form-validate form-horizontal well">
+		<fieldset>
```
Comment (Quy) [inline components/com_mailto/views/mailto/tmpl/default.php:28]: Missing closing `</fieldset>`

Comment (zero-24) [reply components/com_mailto/views/mailto/tmpl/default.php:28]: should be fixed now. Thanks

**Comment by zero-24** [comment]
> In Chrome without captcha, there is a vertical scrollbar.

I have made the frame even a bit more bigger.

**Comment by zero-24** [comment]
@brianteeman can you please approve the language changes (adding the string for Captcha) as the requested changes had been implemented. Thanks.

**Review by brianteeman — APPROVED** [review]

**Comment by Quy** [comment]
Probably beyond the scope of this PR, can we limit the length of subject and sender to curtail spam? Currently, you can submit a paragraph of text/links in the sender field that will be inserted into the body of the email.

**Comment by zero-24** [comment]
> Probably beyond the scope of this PR, can we limit the length of subject and sender to curtail spam? Currently, you can submit a paragraph of text/links in the sender field that will be inserted into the body of the email.

Yes i would call that out of scope but i think we can do that in a following PR when this gets merged. Can you open a issue so we can keep track on that?

File: components/com_mailto/models/mailto.php [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	 *
+	 * @return  array  The default data is an empty array.
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	protected function loadFormData()
+	{
+		$user = JFactory::getUser();
+		$app  = JFactory::getApplication();
+		$data = array();
```
Comment (Quy) [inline components/com_mailto/models/mailto.php]: Remove since it will be reassigned on line 58.

File: components/com_content/helpers/icon.php [hunk 5]
```diff
@@ -73,7 +73,12 @@ public static function email($article, $params, $attribs = array(), $legacy = fa
 		$link     = $base . JRoute::_(ContentHelperRoute::getArticleRoute($article->slug, $article->catid, $article->language), false);
 		$url      = 'index.php?option=com_mailto&tmpl=component&template=' . $template . '&link=' . MailtoHelper::addLink($link);
 
-		$status = 'width=400,height=350,menubar=yes,resizable=yes';
+		$status = 'width=400,height=550,menubar=yes,resizable=yes';
```
Comment (Quy) [inline components/com_content/helpers/icon.php]: Another way to consider:
```
		$height = JFactory::getApplication()->get('captcha', '0') === '0' ? 450 : 550;
		$status = 'width=400,height=' . $height . ',menubar=yes,resizable=yes';
```

File: components/com_mailto/models/mailto.php [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	 */
+	public function getData()
+	{
+		$input = JFactory::getApplication()->input;
+
+		$data['emailto']   = $input->get('emailto', '', 'string');
+		$data['sender']    = $input->get('sender', '', 'string');
+		$data['emailfrom'] = $input->get('emailfrom', '', 'string');
+		$data['subject']   = $input->get('subject', '', 'string');
+		$data['captcha']   = $input->get('captcha', '', 'string');
```
Comment (Quy) [inline components/com_mailto/models/mailto.php]: Remove `$data['captcha']`. Here is a var_dump of `$input`.

```
  ["data":protected]=>
  &array(12) {
    ["task"]=>
    string(4) "send"
    ["emailto"]=>
    string(19) "test@example.com"
    ["sender"]=>
    string(3) "John Doe"
    ["emailfrom"]=>
    string(19) "test@example.com"
    ["subject"]=>
    string(4) "Test"
    ["g-recaptcha-response"]=>
    string(484) "03AJpayVFwI5ojctS5-iOKcPzmEbYu3vMqhLKgpDYwQl00E1I8a2ALO7elw8GU-vMARD7BeQrvFvzoc9VQlJuGXhs7ThW-_lZ5bQzO4ExyKdWNSuVn0Z65qHs58WdVNlulPTi3y0Rr6h9WkJx4dNerSO7e-uRRevtmf264JMUEUUiFFFaXu2nWGuO3J1X3holMHFKxC2dr8kGD_zUPJKPYmJXp2oBGju9H9JH-7fcgRIb-z8qBaovZ4Bf8y-fMO6NNZg1cTnLdR8Agl07RsPUOrZHty4zF4C7YdGer6tgpGbaukGY9uwpfewFIvCuOXWpyaFghlKJE5olc346PTqCaLwrIGhmmJHC_J23KBm96Cllh0w3X7U4YHdmRDKt3xo3nTHRsCd0pM8xW4SiKnFk4CqevseRa59GAQ9Ppu5P5gUwvpxw8rYkPCnerpyu_iSJoinVLgWOJoRefWqzO8mtN_EF2RZ04QZi_0Q"
    ["layout"]=>
    string(7) "default"
    ["option"]=>
    string(10) "com_mailto"
    ["tmpl"]=>
    string(9) "component"
    ["link"]=>
    string(40) "a9738b32019b618d686bd9a69d68c25b1a7b487b"
    ["cebbff4423bc7e9230300bc2d6e82f61"]=>
    string(1) "1"
    ["Itemid"]=>
    NULL
  }
```

**Comment by zero-24** [comment]
Fixed thanks @Quy

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 65b0c3b7e671b48d55f0e3d67b16f6a358684ea8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20265">issues.joomla.org/tracker/joomla-cms/20265</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 65b0c3b7e671b48d55f0e3d67b16f6a358684ea8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20265">issues.joomla.org/tracker/joomla-cms/20265</a>.</sub>

File: components/com_mailto/views/mailto/tmpl/default.php [hunk 7]
```diff
@@ -41,44 +24,36 @@
...
+			<?php foreach ($this->form->getFieldset('') as $field) : ?>
+				<?php if (!$field->hidden) : ?>
+					<div class="control-group">
```
Comment (laoneo) [inline components/com_mailto/views/mailto/tmpl/default.php]: Why do you not use $field->renderField (); ?

**Comment by infograf768** [comment]
Tested fine here. Will mark test OK after the change suggested by @laoneo

File: components/com_mailto/views/mailto/tmpl/default.php (hunk 7 above)
Comment (zero-24) [reply components/com_mailto/views/mailto/tmpl/default.php]: I'm going to implement that when i got back at home today. Thanks!

Comment (laoneo) [reply components/com_mailto/views/mailto/tmpl/default.php]: 👍

Comment (zero-24) [reply components/com_mailto/views/mailto/tmpl/default.php]: fixed with the last commit. Thanks.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 7e70b615302dccc7cd1697cafdc90db07e8ca50e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20265">issues.joomla.org/tracker/joomla-cms/20265</a>.</sub>

**Comment by Quy** [comment]
Ping @infograf768 Thanks.

**Comment by zero-24** [comment]
@infograf768 as the requested change have been implemented can you please do your test so we can move that into 3.8.8? Thanks!

**Comment by zero-24** [comment]
Well this is going to be delayed to 3.8.9. :( @infograf768 can you please mark your test now so we can finally release it in 3.8.9?

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 7e70b615302dccc7cd1697cafdc90db07e8ca50e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20265">issues.joomla.org/tracker/joomla-cms/20265</a>.</sub>

**Comment by infograf768** [comment]
RTC. Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20265">issues.joomla.org/tracker/joomla-cms/20265</a>.</sub>

**Comment by zero-24** [comment]
Thanks
@brianteeman
@mbabker
@ot2sen
@Quy
@infograf768
@laoneo

**Pull request merged**